src/metabase/upload.clj

```diff
@@ -340,3 +348,130 @@
+  [database table]
+  (sync-fields/sync-fields-for-table! database table)
+  (future
+    (sync/sync-table! table)))
+
+(defn- can-upload-error
+  "Returns an ExceptionInfo object if the user cannot upload to the given database and schema. Returns nil otherwise."
+  [db schema-name]
+  (let [driver (driver.u/database->driver db)]
+    (cond
+      (not (public-settings/uploads-enabled))
+      (ex-info (tru "Uploads are not enabled.")
+               {:status-code 422})
+      (premium-features/sandboxed-user?)
+      (ex-info (tru "Uploads are not permitted for sandboxed users.")
+               {:status-code 403})
+      (not (driver/database-supports? driver :uploads nil))
+      (ex-info (tru "Uploads are not supported on {0} databases." (str/capitalize (name driver)))
+               {:status-code 422})
+      (and (str/blank? schema-name)
+           (driver/database-supports? driver :schemas db))
+      (ex-info (tru "A schema has not been set.")
+               {:status-code 422})
+      (not (perms/set-has-full-permissions? @api/*current-user-permissions-set*
+                                            (perms/data-perms-path (u/the-id db) schema-name)))
+      (ex-info (tru "You don''t have permissions to do that.")
+               {:status-code 403})
+      (and (some? schema-name)
+           (not (driver.s/include-schema? db schema-name)))
+      (ex-info (tru "The schema {0} is not syncable." schema-name)
+               {:status-code 422}))))
+
+(defn- check-can-upload
+  "Throws an error if the user cannot upload to the given database and schema."
+  [db schema-name]
+  (when-let [error (can-upload-error db schema-name)]
+    (throw error)))
+
+(defn can-upload?
+  "Returns true if the user can upload to the given database and schema, and false otherwise."
+  [db schema-name]
+  (nil? (can-upload-error db schema-name)))
+
+(defn upload-csv!
+  "Main entry point for CSV uploading.
+
+  What it does:
+  - throws an error if the user cannot upload to the given database and schema (see [[can-upload-error]] for reasons)
+  - throws an error if the user has write permissions to the given collection
+  - detects the schema of the CSV file
+  - inserts the data into a new table with a unique name, along with an extra auto-generated primary key column
+  - syncs and scans the table
+  - creates a model which wraps the table
+
+  Requires that current-user dynamic vars in [[metabase.api.common]] are bound as if by API middleware (this is
+  needed for QP permissions checks).
+  Returns the newly created model. May throw validation, permimissions, or DB errors.
+
+  Args:
+  - `collection-id`: the ID of the collection to create the model in.
+  - `filename`: the name of the file being uploaded.
+  - `file`: the file being uploaded.
+  - `database`: the database to upload to.
+  - `schema-name`: the name of the schema to create the table in (optional).
+  - `table-prefix`: the prefix to use for the table name (optional)."
```

Comment:
could you make this take a `user` argument too? with that we can make this function free of dynamic vars

Reply:
Unfortunately the user dynamic vars need to be set because they're used for permissions checks in the QP when we execute the queries to insert the data. I find this a bit weird and unexpected, but I don't want to make any more changes in this PR.

Reply:
I've left a brief explanation in the docstring https://github.com/metabase/metabase/blob/37a425c52587ca0187562ae7201fd898985c6850/src/metabase/upload.clj#L413-L414